File: benchmarks/ruby/fasta.rb
#!/usr/bin/env ruby
# FASTA benchmark for Ruby

require 'benchmark'

# Constants for the random number generator
IM = 139968
IA = 3877
IC = 29573
SEED = 42

# Define DNA sequences
ALU = "GGCCGGGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCCGAGGCGGGCGGATCACCTGAGGTCAGGAGTTCGAGACCAGCCTGGCCAACATGGTGAAACCCCGTCTCTACTAAAAATACAAAAATTAGCCGGGCGTGGTGGCGCGCGCCTGTAATCCCAGCTACTCGGGAGGCTGAGGCAGGAGAATCGCTTGAACCCGGGAGGCGGAGGTTGCAGTGAGCCGAGATCGCGCCACTGCACTCCAGCCTGGGCGACAGAGCGAGACTCCGTCTCAAAAA"

IUB_PROB = [
    0.27, 0.12, 0.12, 0.27, 0.02,
    0.02, 0.02, 0.02, 0.02, 0.02,
    0.02, 0.02, 0.02, 0.02, 0.02,
]

IUB_CHAR = [
    'a', 'c', 'g', 't', 'B',
    'D', 'H', 'K', 'M', 'N',
    'R', 'S', 'V', 'W', 'Y',
]

HOMO_SAPIENS_PROB = [
    0.3029549426680, 0.1979883004921,
    0.1975473066391, 0.3015094502008,
]

HOMO_SAPIENS_CHAR = [
    'a', 'c', 'g', 't',
]

# Generate a random number
def gen_random(seed)
  value = (seed * IA + IC) % IM
  [value, value.to_f / IM]
end

# Generate a random FASTA sequence
def gen_random_fasta(n, seed, probs, chars)
  length = probs.length
  buffer = ""
  
  n.times do
    seed, r = gen_random(seed)
    c = ' '
    
    length.times do |j|
      if r < probs[j]
        c = chars[j]
        break
      end
      r -= probs[j]
    end
    
    buffer << c
  end
  
  [seed, buffer]
end

# Repeat a sequence until it reaches the required length
def repeat_fasta(n, seq)
  seq_len = seq.length
  buffer = ""
  
  n.times do |i|
    buffer << seq[i % seq_len]
  end
  
  buffer
end

def main
  n = 1_000_000 # Default sequence length
  seed = SEED
  start_time = Time.now
  
  # Write FASTA header and sequence for Homo sapiens Alu
  puts ">ONE Homo sapiens alu"
  alu_seq = repeat_fasta(n, ALU)
  puts alu_seq
  
  # Write FASTA header and random sequence for IUB ambiguity codes
  puts ">TWO IUB ambiguity codes"
  seed, iub_seq = gen_random_fasta(n, seed, IUB_PROB, IUB_CHAR)
  puts iub_seq
  
  # Write FASTA header and random sequence for Homo sapiens frequency
  puts ">THREE Homo sapiens frequency"
  seed, sapiens_seq = gen_random_fasta(n, seed, HOMO_SAPIENS_PROB, HOMO_SAPIENS_CHAR)
  puts sapiens_seq
  
  elapsed = (Time.now - start_time) * 1000
  puts "Time taken: #{elapsed.round} ms"
  
  # Calculate approximate memory usage
  memory_usage = alu_seq.length + iub_seq.length + sapiens_seq.length
  puts "Memory used: #{memory_usage / 1024} KB"
end

main

File: benchmarks/javascript/fasta.js
// FASTA benchmark - generate and write random DNA sequences

// Constants for the random number generator
const IM = 139968;
const IA = 3877;
const IC = 29573;
const SEED = 42;

// Define DNA sequences
const ALU = "GGCCGGGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCCGAGGCGGGCGGATCACCTGAGGTCAGGAGTTCGAGACCAGCCTGGCCAACATGGTGAAACCCCGTCTCTACTAAAAATACAAAAATTAGCCGGGCGTGGTGGCGCGCGCCTGTAATCCCAGCTACTCGGGAGGCTGAGGCAGGAGAATCGCTTGAACCCGGGAGGCGGAGGTTGCAGTGAGCCGAGATCGCGCCACTGCACTCCAGCCTGGGCGACAGAGCGAGACTCCGTCTCAAAAA";

const IUB_PROB = [
    0.27, 0.12, 0.12, 0.27, 0.02,
    0.02, 0.02, 0.02, 0.02, 0.02,
    0.02, 0.02, 0.02, 0.02, 0.02,
];

const IUB_CHAR = [
    'a', 'c', 'g', 't', 'B',
    'D', 'H', 'K', 'M', 'N',
    'R', 'S', 'V', 'W', 'Y',
];

const HOMO_SAPIENS_PROB = [
    0.3029549426680, 0.1979883004921,
    0.1975473066391, 0.3015094502008,
];

const HOMO_SAPIENS_CHAR = [
    'a', 'c', 'g', 't',
];

const BUF_SIZE = 1024 * 1024;

// Generate a random number
function genRandom(seed) {
    const value = (seed[0] * IA + IC) % IM;
    seed[0] = value;
    return value / IM;
}

// Generate a random FASTA sequence
function genRandomFasta(n, seed, probs, chars) {
    const length = probs.length;
    let buffer = '';
    
    for (let i = 0; i < n; i++) {
        let r = genRandom(seed);
        let c = ' ';
        
        for (let j = 0; j < length; j++) {
            if (r < probs[j]) {
                c = chars[j];
                break;
            }
            r -= probs[j];
        }
        
        buffer += c;
    }
    
    return buffer;
}

// Repeat a sequence until it reaches the required length
function repeatFasta(n, seq) {
    const seqLen = seq.length;
    let buffer = '';
    
    for (let i = 0; i < n; i++) {
        buffer += seq[i % seqLen];
    }
    
    return buffer;
}

function main() {
    const n = 1000000; // Default sequence length
    const seed = [SEED];
    const startTime = Date.now();
    
    // Write FASTA header and sequence for Homo sapiens Alu
    console.log(">ONE Homo sapiens alu");
    const aluSeq = repeatFasta(n, ALU);
    console.log(aluSeq);
    
    // Write FASTA header and random sequence for IUB ambiguity codes
    console.log(">TWO IUB ambiguity codes");
    const iubSeq = genRandomFasta(n, seed, IUB_PROB, IUB_CHAR);
    console.log(iubSeq);
    
    // Write FASTA header and random sequence for Homo sapiens frequency
    console.log(">THREE Homo sapiens frequency");
    const sapiensSeq = genRandomFasta(n, seed, HOMO_SAPIENS_PROB, HOMO_SAPIENS_CHAR);
    console.log(sapiensSeq);
    
    const elapsed = Date.now() - startTime;
    console.log(`Time taken: ${elapsed} ms`);
    
    // Get memory stats
    const memoryUsageInMB = process.memoryUsage().heapUsed / 1024 / 1024;
    console.log(`Memory used: ${Math.round(memoryUsageInMB * 1024)} KB`);
}

main();